File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        
        vector<int> left;
        vector<int> right(height.size());
        int maxSeen = 0;
        for (auto num: height) {
            left.push_back(maxSeen);
            maxSeen = max(maxSeen, num);
        }
        
        maxSeen = 0;
        for (int i = height.size() - 1; i >= 0; i--) {
            right[i] = maxSeen;
            maxSeen = max(maxSeen, height[i]);
        }
        
        int result = 0;
        
        for(int i = 0; i < height.size(); i++) {
            int temp = min(left[i], right[i]) - height[i];
            if (temp > 0) result += temp;
        }
        
        return result;
        
        
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:

        leftMax = [0] * len(height)
        rightMax = [0] * len(height)
        
        maxL = 0
        for i,h in enumerate(height):
            leftMax[i] = maxL
            maxL = max(maxL, h)
        
        maxR = 0
        for r in range(len(height) - 1, -1, -1):
            rightMax[r] = maxR
            maxR = max(maxR, height[r])
        
        ans = 0
        for i in range(len(height)):
            temp = min(leftMax[i], rightMax[i]) - height[i]
            if temp > 0:
                ans += temp
        
        return ans
